src/cargo/ops/cargo_new.rs

```diff
@@ -44,38 +57,238 @@ struct CargoNewConfig {
+        return Ok(name);
     }
-    let name = match opts.name {
-        Some(name) => name,
-        None => {
-            let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
-                human(&format!("cannot create a project with a non-unicode name: {:?}",
-                               path.file_name().unwrap()))
-            }));
-            if opts.bin {
-                dir_name
-            } else {
-                let new_name = strip_rust_affixes(dir_name);
-                if new_name != dir_name {
-                    let message = format!(
-                        "note: package will be named `{}`; use --name to override",
-                        new_name);
-                    try!(config.shell().say(&message, BLACK));
-                }
-                new_name
-            }
+    
+    let dir_name = try!(path.file_name().and_then(|s| s.to_str()).chain_error(|| {
+        human(&format!("cannot create a project with a non-unicode name: {:?}",
+                       path.file_name().unwrap()))
+    }));
+    
+    if opts.bin {
+        Ok(dir_name)
+    } else {
+        let new_name = strip_rust_affixes(dir_name);
+        if new_name != dir_name {
+            let message = format!(
+                "note: package will be named `{}`; use --name to override",
+                new_name);
+            try!(config.shell().say(&message, BLACK));
         }
-    };
+        Ok(new_name)
+    }
+}
+
+fn check_name(name: &str) -> CargoResult<()> {
     for c in name.chars() {
         if c.is_alphanumeric() { continue }
         if c == '_' || c == '-' { continue }
         bail!("Invalid character `{}` in crate name: `{}`", c, name)
     }
-    mk(config, &path, name, &opts).chain_error(|| {
+    Ok(())
+}
+
+fn detect_source_paths_and_types(project_path : &Path, 
+                                 project_name: &str, 
+                                 detected_files: &mut Vec<SourceFileInformation>,
+                                 ) -> CargoResult<()> {
+    let path = project_path;
+    let name = project_name;
+    
+    enum H {
+        Bin,
+        Lib,
+        Detect,
+    }
+    
+    struct Test {
+        proposed_path: String,
+        handling: H,
+    }
+        
+    let tests = vec![
+        Test { proposed_path: format!("src/main.rs"),     handling: H::Bin },
+        Test { proposed_path: format!("main.rs"),         handling: H::Bin },
+        Test { proposed_path: format!("src/{}.rs", name), handling: H::Detect },
+        Test { proposed_path: format!("{}.rs", name),     handling: H::Detect },
+        Test { proposed_path: format!("src/lib.rs"),      handling: H::Lib },
+        Test { proposed_path: format!("lib.rs"),          handling: H::Lib },
+    ];
+    
+    for i in tests {
+        let pp = i.proposed_path;
+        
+        // path/pp does not exist or is not a file
+        if ! fs::metadata(&path.join(&pp)).map(|x| x.is_file()).unwrap_or(false) {
```

Comment:
1. Is it really more readable without the space after `!`? Without the space it looks like some auxilary technical `&`.
2. It it important enough to change?